### PR TITLE
[1.x] Generate `AbstractZincFile` during `-sourcepath` workflow

### LINKED ISSUE #1304: MatchError in `ApiPhase.processScalaUnit` with `-sourcepath` compiler option
Setup:
  - small sbt project with two source files for classes A and B
  - set the `-sourcepath` compiler flag
  - change source file B
  - touch (save) the source file A without modifications

Zinc's incremental compilation runs the compiler on the changed source file B. For class A, the compiler finds both a classfile and a source file. Because the [source file has a more recent timestamp it creates a SourcefileLoader](https://github.com/scala/scala/blob/v2.13.12/src/compiler/scala/tools/nsc/symtab/SymbolLoaders.scala#L190-L192) for symbol A, which invokes `compileLate` and creates a plain `BatchSourceFile`.

The API phase crashes because it expects an `xsbti.VirtualFile`: https://github.com/sbt/zinc/blob/v1.9.5/internal/compiler-bridge/src/main/scala/xsbt/API.scala#L62

I'm encountering this when switching branches in scala/scala.

(edit from Dale: switched A and B in the setup, to match the description below.)

### PR DESCRIPTION
**Pull request by Friendseeker**
This PR fixes #1304, #963

### Issue

https://github.com/sbt/zinc/blob/5eb0de1fdfd921f0d1b6a9302fe42593e7e27922/internal/compiler-bridge/src/main/scala/xsbt/API.scala#L62

In certain codepath, `unit.source.file` is a `PlainFile`, which causes a `MatchError`.

### Fix

When `unit.source` is built, convert `unit.source.file` to `AbstractZincFile`.

### Validating the fix

The PR includes reproduction scripted test by [lrytz](https://github.com/lrytz). The test passes after the fix.

### CODE REVIEW

File: internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala [hunk 1]
```diff
@@ -78,6 +78,16 @@ sealed class ZincCompiler(settings: Settings, dreporter: DelegatingReporter, out
     extends CallbackGlobal(settings, dreporter, output)
     with ZincGlobalCompat {
 
+  override def getSourceFile(f: AbstractFile): BatchSourceFile = {
+    val file = f match {
+      case plainFile: PlainFile =>
+        val vf = callback.asInstanceOf[AnalysisCallback3].toVirtualFile(plainFile.file.toPath)
+        new ZincPlainFile(vf.asInstanceOf[PathBasedFile])
```
Comment (Friendseeker) [inline internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala]: The conversion logic is... hacky.

I would appreciate any feedback suggesting a better solution.

File: internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala [hunk 2]
```diff
@@ -78,6 +79,19 @@ sealed class ZincCompiler(settings: Settings, dreporter: DelegatingReporter, out
...
     with ZincGlobalCompat {
 
+  override def getSourceFile(f: AbstractFile): BatchSourceFile = {
+    val file = (f, callback) match {
+      case (plainFile: PlainFile, callback3: AnalysisCallback3) =>
+        val virtualFile = callback3.toVirtualFile(plainFile.file.toPath)
+        virtualFile match {
+          case underlying: PathBasedFile => new ZincPlainFile(underlying)
+          case _                         => f
+        }
```
Comment (dwijnand) [inline internal/compiler-bridge/src/main/scala/xsbt/CallbackGlobal.scala]: I think this is `AbstractZincFile(callback3.toVirtualFile(plainFile.file.toPath))`, btw.

**Review by dwijnand — CHANGES_REQUESTED** [review]
See https://github.com/sbt/zinc/issues/1304#issuecomment-1875188256

**Review by eed3si9n — APPROVED** [review]

**Comment by Friendseeker** [comment]
@eed3si9n While the PR is in `Changes requested` status, would it be possible for you to dismiss the stale review and force merge it if everything looks okay?

My concern is that since this PR involves `AnalysisCallback3` and https://github.com/sbt/zinc/pull/1306 is already merged in, if a Zinc release is made before the PR is merged, we would need to introduce `AnalysisCallback4`.

**Comment by eed3si9n** [comment]
Yea. Now that AnalysisCallback is AnalysisCallback3, we should just land this.

**Pull request merged**